repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2Filter.scala

```diff
@@ -0,0 +1,585 @@
+                authorizeTenant(tenantFromUri, validToken) flatMap { _ =>
+                  lazy val endpoints = getEndpoints(getAdminToken, authToken, validToken) // Prevents making call if its not needed
+                  addCatalogHeader(endpoints) flatMap { _ =>
+                    authorizeEndpoints(validToken, endpoints) flatMap { _ =>
+                      addGroupsHeader(getGroups(getAdminToken, authToken, validToken))
+                    }
+                  }
+                }
+              }
+            }
+
+          processingResult match {
+            case Success(_) => Pass
+            case Failure(e: MissingAuthTokenException) => Reject(SC_UNAUTHORIZED, Some(e.getMessage))
+            case Failure(e: InvalidTokenException) => Reject(SC_UNAUTHORIZED, Some(e.getMessage))
+            case Failure(e: InvalidTenantException) => Reject(SC_UNAUTHORIZED, Some(e.getMessage))
+            case Failure(e: UnparseableTenantException) => Reject(SC_UNAUTHORIZED, Some(e.getMessage))
+            case Failure(e: IdentityCommunicationException) => Reject(SC_BAD_GATEWAY, Some(e.getMessage))
+            case Failure(e: UnauthorizedEndpointException) => Reject(SC_FORBIDDEN, Some(e.getMessage))
+            case Failure(e: OverLimitException) =>
+              response.addHeader(HttpHeaders.RETRY_AFTER, e.retryAfter)
+              Reject(HttpServletResponse.SC_SERVICE_UNAVAILABLE, Some(e.getMessage))
+            case Failure(e) => Reject(SC_INTERNAL_SERVER_ERROR, Some(e.getMessage))
+          }
+        }
+
+      filterResult match {
+        case Pass =>
+          logger.trace("Processing completed, passing to next filter or service")
+          addIdentityStatusHeader(confirmed = true)
+          chain.doFilter(request, response)
+        case Reject(statusCode, message) =>
+          Option(config.getDelegating) match {
+            case Some(delegating) =>
+              logger.debug(s"Delegating with status $statusCode caused by: ${message.getOrElse("unspecified")}")
+
+              val delegationHeaders = buildDelegationHeaders(statusCode,
+                "keystone-v2",
+                message.getOrElse("Failure in the Keystone v2 filter").replace("\n", " "),
+                delegating.getQuality)
+
+              addIdentityStatusHeader(confirmed = false)
+              delegationHeaders foreach { case (key, values) =>
+                values foreach { value =>
+                  request.addHeader(key, value)
+                }
+              }
+
+              chain.doFilter(request, response)
+
+              logger.trace(s"Processing response with status code: $statusCode")
+
+              val wwwAuthenticateHeader = response.getHeader(CommonHttpHeader.WWW_AUTHENTICATE)
+
+              response.getStatus match {
+                case HttpServletResponse.SC_UNAUTHORIZED | HttpServletResponse.SC_FORBIDDEN =>
+                  if (DELEGATED.equalsIgnoreCase(wwwAuthenticateHeader.trim)) {
```

Comment:
Oh, wait, actually... so if I remove the conditional, it may be unsafe to set the www-authenticate header as that may remove an existing valid value. If i always add our value, though, then we may pass the "Delegated" value back to the user when delegation should really be transparent.

Still, not a big deal, but I'll wait for confirmation before removing this.